### 接口错误码整理之异常处理

#### 背景

最近问题反馈群里有用户反馈说：在商旅预订酒店过程中会弹框“正在计算，请稍后”。用户一脸蒙逼，什么鬼。。。问题反馈到资金这边，通过log发现原来是企业金额不足。虽然接口返回对象中有把该错误信息带过去，但是很明显，酒店那边没有把商旅的错误信息展示出来，同时也发现商旅这边错误信息提示得不够明确，借这个机会整理下商旅冻结接口的错误信息和推动酒店展示商旅错误文案。

#### Action 1

跟产品说了想法后，产品觉得很赞，随后确定了两件事情：1）我负责整理冻结时可能出现的错误文案；2）联系酒店，机票的PD，经沟通，最终达成的一致是：商旅负责吐文案，吐什么就展示什么，然后协调开发资源。

#### Action 2

把可能出现的错误文案整理给PD，并修改成用户能理解的文案后，发给酒店，机票PD过目，没问题，商旅走透传，各类目展示。

#### Action 3

商旅的冻结接口拿不到底层返回的具体错误码，what the fuck??? 看代码。

调用链路：支付(HSF接口) -> 支付(Manager) -> 资金

支付(Manager)加了分布式锁和Spring事务注解，资金也加了分布式锁和Spring事务注解，事务传播行为是REQUIRED，分布式锁使用Spring环绕通知，且优先级最高，在所有切面前先执行。

这个流程的结构就像这样：

	分布式锁层
	try {
		// 事务层
		try {
			// A方法
			A();
		} catch(Exception e) {
			throw UnexpectedRollbackException(e);
		}
	} catch(CreditBizException e) { // 自定义的异常
		throw e;
	} catch(Throwable e) {
		throw new CreditBizException(e);
	} finally {
		// 释放锁操作
	}
	
	@Transactional
	public Result A() {
		try {
			B();
		} catch(CreditBizException e) {
			// do something
			return Result;
		}
	}
	
	@Transactional
	public Result B() throw CreditBizException {
		throw new CreditBizException(errorCode, errorMsg);
	}

A方法相当于支付的Manager层，B方法相当于资金层。A，B方法在同一个事务中，且A方法中try catch了B方法抛出的异常。

假如B方法中抛出了余额不足的CreditBizException异常，此时事务会被标记为需要roll-back. 在A方法中，由于catch了B方法的异常，处理逻辑是return 返回值，当A方法尝试提交本次数据库变更，而由于该事务已经被标记为roll-back，那么本次提交会导致抛出UnexpectedRollbackException异常，此时在分布式锁的catch中，拿到的errorMsg已经不是我们想要的错误文案了。

怎么改进了？

- A方法中需要catch B方法的异常吗
- 如何防止UnexpectedRollbackException异常抛到支付(HSF接口)层

#### Action 4

A方法中需要catch B方法的异常吗？ 需要，因为有些错误的文案，比如DB异常，一些未知错误等底层返回的错误，不能直接吐给调用方，需要统一转换为系统异常，且我们在这一层对错误日志做了埋点。

如何防止UnexpectedRollbackException异常抛到支付(HSF接口)层？ 在A方法中，当catch到B的异常时，应该把B的errorCode和errorMsg封装后往上抛，即：

	@Transactional
	public Result A() {
		try {
			B();
		} catch(CreditBizException e) {
			String errCode = e.getErrCode();
			String errMsg = e.getErrMsg();
			throw new CreditBizException(errCode, errMsg);
		}
	}

这时，A中虽然catch了这个异常，但不会尝试提交它的变更，会把B的异常信息抛出来，在分布式锁的catch中拿到这个异常，然后支付HSF层catch到这个异常后，就是底层抛出来的错误文案了。

#### Action 5

商旅端终于能把错误文案吐出来了，心情舒服多了。
